Switch git to inline callbacks (update #1793)

djmitche opened the pull request:
This updates #1793 by merging to master and resolving some conflicts and quite a bit of bitrot.

Issues:
 * `_updateSubmodule` and `_cleanSubmodule` were asynchronous but called from `clean` without `yield`
 * `copy` called `mode_incremental` and expected a return code, but that method returns nothing
 * `startVC` called the parent `patch` method without supplying a value for `_`.

@vlovich what do you think?

Stibbons commented on master/buildbot/steps/source/git.py:
```diff
@@ -222,14 +217,26 @@ def mode_incremental(self, _):
+
+        rc = yield self._fetchOrFallback()
+        if rc != RC_SUCCESS:
+            raise buildstep.BuildStepFailed
```
What about refactor in a loop:

for clean_meth in ['_fetchOrFallback', '_syncSubmodule', ...]:
    rc = yield getattr(self, clean_meth)
    if rc != RC_SUCCESS:
        raise buildstep.BuildStepFailed

tardyp replied:
I don't like the loop think, but I dont like the lots of if either.

In repo, there is a abandonOnFailure in _dovccmd likes. I think this is a better method for handling this problem.

tardyp commented on master/buildbot/steps/source/git.py:
```diff
@@ -251,46 +258,35 @@ def fresh(self):
+                            timeout=self.timeout)
 
         old_workdir = self.workdir
         self.workdir = self.srcdir
```
smelly

djmitche replied:
Yeah, the idea is to re-use the full checkout code but pointed at the srcdir.

tardyp commented on master/buildbot/steps/source/git.py:
```diff
@@ -251,46 +258,35 @@ def fresh(self):
+                                           'logEnviron': self.logEnviron,
+                                           'timeout': self.timeout, })
+            cmd.useLog(self.stdio_log, False)
+            yield self.runCommand(cmd)
```
wouldn't it better to say here

    yield self.runCommand(cmd, workdir= self.srcdir)

tardyp commented:
Thanks for working on this.  The new code looks so much better!

I would expect 90% diff hit is a little bit low for a source step. especially that you found issues with the old code.

djmitche commented:
Well, I just rebased it and fixed some test failures.  I'm really not sure where those test failures came from, but I think it's just bitrot - the original patch passed its tests before rebase.

As for the suggested changes, I'll leave that to @vlovich

tardyp commented:
ok, lets merge it then :+1:
